### DISCUSSION TITLE
Beelink U59 + Ubuntu 24.04 LTS

### DISCUSSION THREAD

**Discussion by Cold-Lemonade** [OpenSurv/OpenSurv#27] [Hardware + Operating system]
OpenSurv is very cool. I've got it to display the rtsp streams from 9 cameras in a 3x3 gird using a Beelink U59 (Intel N5105 with 16 GB ram). However, the streams were not synchronized as some streams were a few seconds ahead and some a few seconds behind. I understand that vlc has a ```--sout-smem-time-sync``` option. Should that be passed through using ```freeform_advanced_mpv_options:```? I think that would delay all the streams to match the slowest one to start and so I wouldn't be seeing "live" images once OpenSurv started showing the camera feeds.

The cameras are connected to the network over cat6 and their rtsp streams are being restreamed via Frigate NVR. But mini pc running OpenSurv connects to the network over wifi. If I connect the mini pc to the network over cat6 as well, would that reduce the latency?

**Comment by SvenVD** [comment]
Try to lower latency with https://github.com/OpenSurv/OpenSurv/blob/e904ccc128fe56990242b801af3437302a5d79d1/README.md?plain=1#L133 if they all have low latency then they should be quite in sync with each other

**Reply by SvenVD** [reply]
Also wired connection is much more stable and predictable then a wifi connection

**Reply by Cold-Lemonade** [reply]
@SvenVD Thank you. Adding ```--profile=low-latency --untimed``` got the 9 rtsp streams almost perfectly synchronized. The caveat is that these are all low res streams around 704x480 resolution at 5 or 8 frames per second, which OpenSurv was scaling down to 640x360 (because each one was taking up 1/9 of a 1080p monitor). When I use the high res streams which are a mix of 2MP, 4MP, and 8MP running at 15-20 frames per second, the mini pc doesn't do as well. I think it overwhelms this low power computer.

The Beeline U59 seems to handle 4 high res rtsp streams at 15-20 fps without any issue. But any more than that and it struggles. Eventually camera feeds crash and it reloads them. I think the solution is to lower the fps. Is there a way to tell OpenSurv to, for example, use 5 fps?

**Reply by SvenVD** [reply]
Ok great news! but why should you use the high res streams when they only take 1/9th of the monitor, I suspect the downscaling is the performance hit. 

To lower fps the best place would be in the streams settings themselves on the source, not in OpenSurv

**Reply by Cold-Lemonade** [reply]
Right now, I am testing this out on a small portable monitor with 1080p resolution, but my plan is to deploy it on a 4k display. I am thinking about showing 6 cameras feeds with one 4 times larger than the other 5, like you do in the demo.

I think mpv can change the fps using this option ```vf=fps=<float>```. At least, that's what I read here: [https://github.com/mpvnet-player/mpv.net/issues/295#issuecomment-889915044](https://github.com/mpvnet-player/mpv.net/issues/295#issuecomment-889915044) I will try that out later today.

**Reply by SvenVD** [reply]
You probably could use the mpv options, however, if possible, decreasing fps on the source of the stream itself is more efficient in general.

**Reply by Cold-Lemonade** [reply]
Here is a quick update... I am running my ip cameras at their maximum resolution (2688x1520) and I have successfully configured OpenSurv to display them. I have constructed two screens and OpenSurv switches between them every 5 minutes. I am using a 4k monitor with resolution 3840x2160. OpenSurv rescales five rtsp feeds to 1280x720 and a sixth to 2560x1440. 

![left](https://github.com/user-attachments/assets/2a0705d5-356b-4a3f-95da-dd96d987cf92)

![right](https://github.com/user-attachments/assets/f10e363a-01a9-4da9-9f3d-d143f603ac51)

To make this possible I have had to drop the frames per second to 15, but I do this in OpenSurv. So the option line in my OpenSurv config is the following for each camera:

```
freeform_advanced_mpv_options: "--rtsp-transport=tcp --profile=low-latency --untimed --vf=fps=15"
```

I have no issues with lag even though I am connecting to my network over wifi. The only thing seems to be that OpenSurv takes about one second to display each camera feed when it switches between the two screens. Is there a way to load the unseen screen in the background to avoid this? Average cpu load is between 2.5 and 3.0 on this 4 core mini pc and I am using less than 2.6 GB of ram out of 16 GB.

I understand that it may be more efficient to change the resolution of the rtsp streams in the cameras themselves. However, I monitor and record these rtsp streams using Frigate NVR and want to avoid interfering with the quality of the recordings by lowering their resolutions. Also, I only run OpenSurv when I am in my study where the 4k monitor hangs on a wall. If I am not in my study, the Beelink U59 is off. Given that OpenSurv on the Beelink U59 can handle the rescaling I need without issue, I am happy to maintain the current setup unless I see it become unstable. So far, no problems but I have only had it running for 30 minutes or so.

**Reply by SvenVD** [reply]
Thanks for the extensive reporting back. 

Normally OpenSurv should [cache ](https://github.com/OpenSurv/OpenSurv/blob/e904ccc128fe56990242b801af3437302a5d79d1/surveillance/etc/general.yml#L) next screen for it to be shown more quickly.

If the beelink can handle it, there is no problem. But for reference, most IP cameras typically offer two streams: a main stream and a substream, which can be configured independently. You could set Frigate to use the main streams and OpenSurv to use the substreams.
